modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm

```diff
@@ -2727,6 +2727,106 @@ sub _build_circular_slice_cache {
+
+  Args       : none
+  Example    : $self->create_chromosome_alias();
+  Description: create chromosome alias in coordinate system with karyotype attributes
+  Returntype : Bio::EnsEMBL::CoordSystem, or none 
+  Exceptions : None
+  Caller     : general
+  Status     : in testing
+
+=cut
+
+sub _create_chromosome_alias {
+  my $self = shift;
+  my $csa  = $self->db->get_CoordSystemAdaptor();
+
+  # to store reformatted db query results
+  my %karyotype_seq_regions;
+  
+  my $karyotype_rank_string = "karyotype_rank";
+
+  # to store coord_system_id to alias 
+  my $cs_id;
+
+  # check whether seq region cache exists, otherwise run database query
+  if ( $self->{'karyotype_cache'} ) {
+    %karyotype_seq_regions = $self->{'karyotype_cache'};
+  } else {
+    # cannot find a coordssystem called chromosome
+    # look through attribs to find seq_regions with karyotype
+
+    my $sth =
+      $self->prepare( "SELECT sr.seq_region_id, sr.name, sr.coord_system_id, sr.length, a.code, cs.rank "
+                    . "FROM seq_region sr, seq_region_attrib sra, attrib_type a, coord_system cs "
+                    . "WHERE sr.seq_region_id = sra.seq_region_id "
+                    . "AND a.attrib_type_id = sra.attrib_type_id "
+                    . "AND cs.coord_system_id = sr.coord_system_id "
+                    . "AND a.code = ?" );
+
+    $sth->bind_param( 1, $karyotype_rank_string );
+    $sth->execute();
+
+    # fetch SQL results and identify coord_system_id(s) that
+    # has(have) karyotype attribs
+    my %cs_id_rank;
+    while ( my $hashref = $sth->fetchrow_hashref() ) {
+      my $seq_region_id   = $hashref->{ seq_region_id };
+      my $seq_region_name = $hashref->{ name };
+      my $coord_system_id = $hashref->{ coord_system_id };
+      my $length          = $hashref->{ length };
+      my $code            = $hashref->{ code };
+      my $rank            = $hashref->{ rank };
+
+      # account for possibility of multiple coord_system_ids returned from db query
+      $karyotype_seq_regions{ $seq_region_name }{ $coord_system_id } = {
+        code            => $code,
+        length          => $length,
+        seq_region_id   => $seq_region_id,
+        rank            => $rank
+      };
+
+      # hash to help decide which cs id to use, if multiple
+      $cs_id_rank{ $coord_system_id } = $rank;
+    }
+    $sth->finish();
+    my $cs_id_count = keys %cs_id_rank;
+
+    # if number of coordSystem ids retrieved is one, simply use this coordSystem id
+    # otherwise, choose the coordSystem with the best-rank (i.e. lowest number)
+    if ( $cs_id_count == 1 ) {
+      $cs_id = (keys %cs_id_rank)[0]; # get only key name in hash
+      print "using cs dbid: $cs_id\n";
+    } else {
+      foreach my $id (sort { $cs_id_rank{$a} <=> $cs_id_rank{$b} } keys %cs_id_rank) {
+        $cs_id = $id;
+        print "$id, $cs_id_rank{$id}\n";
+        last; # exit loop after getting lowest-ranked coordSystem id
+      }
+    }
```

Comment:
@ens-bwalts I've altered the structure of `%karyotype_seq_regions` so that it no longer assumes that there will only be a single coordSystem with karyotype attributes. The `rank` attribute is also now retrieved from the database, so that in the case of multiple coordSystems with karyotype attributes, the coordSystem with the best rank (i.e. lowest rank number) is chosen and used to create the alias in.

Reply:
@ens-bwalts I've also written a separate simple perl script to loop through all species' databases and look for such an example. I haven't been able to find such an example, but of course there could be an example at a future date so I'm thinking of how to create an appropriate test.

Reply:
Added test in with commit https://github.com/Ensembl/ensembl/pull/532/commits/81c3e573653825f3c2d4624ccdcd5c6d8f5b496d